library/cipher.c

```diff
@@ -215,8 +215,9 @@ int mbedtls_cipher_set_iv( mbedtls_cipher_context_t *ctx,
                    const unsigned char *iv, size_t iv_len )
 {
     size_t actual_iv_size;
-
-    if( NULL == ctx || NULL == ctx->cipher_info || NULL == iv )
+    if( NULL == ctx || NULL == ctx->cipher_info )
+        return( MBEDTLS_ERR_CIPHER_BAD_INPUT_DATA );
+    else if( NULL == iv && iv_len != 0  )
```

Comment:
I am not sure this really addresses the problem reported. This check just ensures that iv is NULL whenever the len is 0, but it changes the behaviour of the function as it is now possible to configure 0 length IV for any mode of operation, not just ECB...

Reply:
the documentation states:

> * \note                Some ciphers don't use IVs nor NONCE. For these
 *                      ciphers, use iv = NULL and iv_len = 0.

I think that this check is exactly what's in the documentation. We check that if the IV is NULL, then the length should be zero, otherwise it will fail, as before.
The funcion before supported your use case of IV with length zero, but then it failed later in the flow:
~~~
        actual_iv_size = ctx->cipher_info->iv_size;

        /* avoid reading past the end of input buffer */
        if( actual_iv_size > iv_len )
            return( MBEDTLS_ERR_CIPHER_BAD_INPUT_DATA );
~~~ 
In fact, our test suite tests this [use case](https://github.com/ARMmbed/mbedtls/blob/72ea31b026e1fc61b01662474aa5125817b968bc/tests/suites/test_suite_cipher.function#L120):
~~~
    /* IV too small */
    TEST_ASSERT( mbedtls_cipher_set_iv( &ctx, iv, 0 )
                 == MBEDTLS_ERR_CIPHER_BAD_INPUT_DATA );
~~~

Reply:
Ah ok, I see what you mean. In that case, I wonder whether we should check that if there is no IV and the len is 0 then a suitable cipher mode is chosen.

Reply:
I think that since it's in our documentation, it's up to the developer to determine the correct cipher mode.